### PR TITLE
[TOPI] Average Pool2D Bug.

### PR DESCRIPTION
**Pull request by anijain2305**
Issue - https://github.com/dmlc/tvm/issues/3581

@yzhliu @tqchen @FrozenGene @shoubhik @rankyung

### CODE REVIEW

**Comment by FrozenGene** [comment]
As https://github.com/dmlc/tvm/issues/3581 discussed, TFLite's tvm.sum type is UINT16 (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L2976), not int32. Could we consider this corner case? @tqchen 's said, we could compose this op to handle it, if so, could we add one test to cover?

**Comment by anijain2305** [comment]
@FrozenGene Upcasting to uint16 as done in TFLite is little out of context in this PR.

For now, I have added one more test case checking `uint16` datatype to ensure that accumulation and division happens at the right time. I think that should be good enough for now.

Let me know your comments on the CPP code changes. I am not very familiar with this section of code. Lets try to get this in quickly because this is a functional bug which will certainly affect automatic quantization accuracy as well.

File: topi/include/topi/nn/pooling.h [hunk 1]
```diff
@@ -129,19 +129,26 @@ inline Tensor pool_impl(const Tensor& x,
...
+    // TVM compute for summing the pooling window.
+    auto pool_sum = tvm::compute(out_shape,
+    [&](const Array<Var>& output) {
       Array<Expr> indices;
       for (const Var& var : output) indices.push_back(var);
       indices.Set(height_axis, output[height_axis] * stride_height + dheight);
       indices.Set(width_axis, output[width_axis] * stride_width + dwidth);
-      return tvm::sum(temp(indices) / divide_factor, { dheight, dwidth });
-    };
+      return tvm::sum(temp(indices), { dheight, dwidth });
```
Comment (FrozenGene) [inline topi/include/topi/nn/pooling.h]: If we could leave one comment of TODO, it is better. Remind us there is one potential issue we should care in the future parsing TFLite model.

Comment (yzhliu) [reply topi/include/topi/nn/pooling.h]: @FrozenGene If you mean int8<->uint8 casting for tflite, I think the TODO can be put into tflite frontend instead. The bugfix here in topi is complete IMO. what do you think?

Comment (FrozenGene) [reply topi/include/topi/nn/pooling.h]: I mean the `tvm.sum` type is `uint16`, not `int32`.

Comment (anijain2305) [reply topi/include/topi/nn/pooling.h]: For TFLite adaptation, we can `cast` the input to `uint16` and then pass the casted input to `avg_pool2d`. This is purely the job of framework parser. It does not need any code in the TOPI layer.

Comment (yzhliu) [reply topi/include/topi/nn/pooling.h]: yes my understanding is `avg_pool2d(int_8_value.astype("uint16")).astype("int8")` will be in frontend, right?

**Comment by tqchen** [comment]
Thanks @FrozenGene @anijain2305 @yzhliu .

File: topi/include/topi/nn/pooling.h (hunk 1 above)
Comment (FrozenGene) [reply topi/include/topi/nn/pooling.h]: I understand the way you want to do. Even the output dtype is "int32", you also want to do `avg_pool2d(int_8_value.astype("int32")).astype("int8")` in frontend.

Comment (FrozenGene) [reply topi/include/topi/nn/pooling.h]: If so, the way is ok.
